Js2 authentication

RamtinMoshtagh opened the pull request:
Authentication is a success now. Please review my code and share thoughts/improvements.

RamtinMoshtagh committed:
firstFeedback

RamtinMoshtagh committed:
Your descriptive commit message

RamtinMoshtagh committed:
Merge branch 'main' of https://github.com/RamtinMoshtagh/css-frameworks-ca

RamtinMoshtagh committed:
changed based on review

RamtinMoshtagh committed:
adjusting to meet requirments

RamtinMoshtagh committed:
more adjustments to meet the criteria

RamtinMoshtagh committed:
more adjustments to meet the criteria

RamtinMoshtagh committed:
Update README.md

RamtinMoshtagh committed:
setting up authentication

RamtinMoshtagh committed:
authentication API request working

RamtinMoshtagh commented:
@sulenchy

RamtinMoshtagh committed:
some adjustments to see if error still there

RamtinMoshtagh committed:
debugging errors - success

sulenchy commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
 .DS_Store
 dist/
+node_modules/
+JavaScript2/
```
Why ignoring this? Why not delete it entirely?

RamtinMoshtagh replied:
true, removed.

sulenchy commented on js/auth.js:
```diff
@@ -0,0 +1,85 @@
+const API_BASE_URL = 'https://api.noroff.dev/api/v1';
```
I will advice you extract this to utility file because you will need this in other files

RamtinMoshtagh replied:
yes, very true. done!

sulenchy commented on src/scss/index.css:
```diff
@@ -0,0 +1,134 @@
+body {
```
I thought you are to use scss and not css. Why the switch?

sulenchy commented on src/scss/index.css:
```diff
@@ -0,0 +1,134 @@
+body {
```
I thought you are to use SCSS and then transpile it to CSS with SASS npm package. Why the switch?

RamtinMoshtagh replied:
hmm, if you look at index.scss there you can see that is has compiled scss to css?

sulenchy reviewed (commented):
@RamtinMoshtagh  I left some feedback for you above. Please see to it.

sulenchy commented on index.html:
```diff
@@ -0,0 +1,77 @@
+                        <div class="form-text">Password must be at least 8 characters long.</div>
+                    </div>
+                    <div id="error-message" class="text-danger text-center"></div>                    
+                    <button type="submit" class="btn w-100">Login and Submit</button>
```
Is this login or register?

RamtinMoshtagh replied:
yes, true its confusing, fixing it now

RamtinMoshtagh committed:
Changes after PR

RamtinMoshtagh committed:
filtering posts

RamtinMoshtagh committed:
fetching post by IDs for detail page

RamtinMoshtagh committed:
adjustments for req criteria

RamtinMoshtagh committed:
debug and adjustment

to meet the requirments and criteria

RamtinMoshtagh committed:
debug createPost function

uploading img with post

RamtinMoshtagh committed:
arranging all my js code

clean, not repeat, structure

RamtinMoshtagh committed:
ES& modules

organize the code and following the criteria

RamtinMoshtagh committed:
debug

RamtinMoshtagh committed:
debugging and testing

testing login,register,fetchposts,delete,edit,redmore,tags,filtertags and making that its error-free

RamtinMoshtagh commented:
@sulenchy coach, i have now completed my code for the CA, and need a review on it to adjust and meet criteria.

sulenchy commented on js/auth.js:
```diff
@@ -0,0 +1,139 @@
+
+        if (!response.ok) {
+            const errorData = await response.json();
+            if (errorData.errors && errorData.errors[0].message === 'Profile already exists') {
```
Nice initiative for handling this error and presenting a directive for the user on the next step.

RamtinMoshtagh replied:
Thanks coach

sulenchy commented on js/auth.js:
```diff
@@ -0,0 +1,139 @@
+        });
+
+        if (!response.ok) {
+            throw new Error('Login failed');
```
You can provide the user an info on what to do next when he/she fails to login successfully. Here is an example
`Login failed. Please, register as a new user`

RamtinMoshtagh replied:
Yes, agreed! Done

RamtinMoshtagh replied:
Thanks coach

sulenchy commented on js/auth.js:
```diff
@@ -0,0 +1,139 @@
+            errorMessageDiv.textContent = 'User registered successfully!'; // Display a success message
+            errorMessageDiv.classList.remove('text-danger');
+            errorMessageDiv.classList.add('text-success');
+            window.location.href = '/profile/index.html';
```
Do you want to log in the user when he or she registers successfully? If yes, then how do you generate the token? According to the API doc., the token is the authentic means of identifying an auth user; without it, the user is not authenticated.

What I am saying essentially is that the user flow is incorrect. You should not take the user to the profile page after a successful registration.

RamtinMoshtagh replied:
Very true, thanks coach, i modified it like this:
- Separated Registration and Authentication.
- Handle Token: to automatically log the user in after registration, once the login is successful, the server should return an authentication token. Token is stored and used for subsequent authenticated requests.
- Redirect to Profile: Only after obtaining the token, redirecting the user to the profile page.

sulenchy commented on js/auth.js:
```diff
@@ -0,0 +1,139 @@
+        localStorage.setItem('accessToken', data.accessToken);
+        
+         // Redirect to the profile or home page
+         window.location.href = '/profile/index.html';
```
This looks good

RamtinMoshtagh replied:
thanks

RamtinMoshtagh replied:
Thanks coach

sulenchy commented on js/postDetail.js:
```diff
@@ -0,0 +1,54 @@
+document.addEventListener('DOMContentLoaded', () => {
+    const urlParams = new URLSearchParams(window.location.search);
+    const postId = urlParams.get('id');
+    if (postId) {
```
What happens when this return false? One thing you could do here is to send a message informing about what has happened

RamtinMoshtagh replied:
Thanks coach, with these changes, when a user visits the page without a valid postId, the Bootstrap modal will pop up, displaying the error message. The user can then close the modal by clicking the "Close" button or anywhere outside the modal.

sulenchy reviewed (commented):
@RamtinMoshtagh Just dropped some comments for you. Check it out. That aside, you need to get feedback from any other person. I suggest you share your with your colleagues in the peer review channel.

RamtinMoshtagh committed:
Update README.md

update the img and included js under "built with"

RamtinMoshtagh committed:
changes after coach review

RamtinMoshtagh committed:
Merge branch 'js2' of https://github.com/RamtinMoshtagh/css-frameworks-ca into js2

RamtinMoshtagh commented:
@sulenchy i have now modified and made changes. I have also asked for feedback in peer review channel. Im getting ready to deliver. what are your thoughts?

RamtinMoshtagh committed:
modification based on feedback

RamtinMoshtagh committed:
Last modification and cleaning

Cleaning extra spaces, comments, console.logs etc

RamtinMoshtagh committed:
Update README.md

img new

RamtinMoshtagh committed:
Update README.md

final img
